### PR TITLE
Sorted Songs in Songs list

### PR DESCRIPTION
**Pull request by AjaysinhRajput2802**
Solved issue #54

### CODE REVIEW

**Review by sanket143 — CHANGES_REQUESTED** [review]
There a lot of redundant file updates in this and other PR too, keep the changes focused only on the issues the PR is solving.

File: scripts/setup.js [hunk 1]
```diff
@@ -35,7 +50,13 @@ document.addEventListener("keypress",function(){
 
       finder.on('file', (file) => {
         if (path.extname(file) == ".mp3") {
-          state.allFiles.push(file)
+          if(!state.nowplaying.instance){
```
[the remark this answers is not in the document]
Comment (AjaysinhRajput2802) [reply scripts/setup.js]: made changes which are requested by you

File: scripts/setup.js [hunk 2]
```diff
@@ -50,17 +36,16 @@ document.addEventListener("keypress", function(){
...
+          state.allFiles.push(file)
+          if(!state.nowplaying.instance) // For first time nowplaying.instance not exists so we check for it
+          {
+            state.load(state.allFiles[0]) // load first song
           }
         }
       })
 
       finder.on('end', async () => {
+        state.allFiles.sort()
```
Comment (sanket143) [inline scripts/setup.js:48]: `allFiles` is a list of file paths and it's being sorted by file names when it is supposed be based on the song name.

**Review by sanket143 — APPROVED** [review]

File: scripts/setup.js [hunk 3]
```diff
@@ -36,10 +37,15 @@ document.addEventListener("keypress",function(){
...
           state.allFiles.push(file)
+          if(!state.nowplaying.instance) // For first time nowplaying.instance not exists so we check for it
+          {
+            state.load(state.allFiles[0]) // load first song
+          }
         }
       })
 
       finder.on('end', async () => {
+        state.allFiles.sort()
```
Comment (sanket143) [inline scripts/setup.js:48]: `allFiles` is a list of file paths and it's being sorted by file names when it is supposed be based on the song name.

**Review by sanket143 — CHANGES_REQUESTED** [review]

File: scripts/setup.js (hunk 3 above)
Comment (AjaysinhRajput2802) [reply scripts/setup.js:48]: @sanket143 if we sort by path then songs were sorted by name automatically.

Comment (sanket143) [reply scripts/setup.js]: they won't get sorted by song names, they'll get sorted by filenames or their full path to be precise.

Comment (AjaysinhRajput2802) [reply scripts/setup.js]: @sanket143 made changes which are requested by you

**Comment by AjaysinhRajput2802** [comment]
@sanket143 @janvig please review this PR.

File: scripts/setup.js [hunk 4]
```diff
@@ -40,6 +40,9 @@ document.addEventListener("keypress",function(){
       })
 
       finder.on('end', async () => {
+        state.allFiles.sort(async (a, b)=>{
+          return state.songsMap[a].title.toLowerCase() < state.songsMap[b].title.toLowerCase() ? -1 : 1
+        })
```
Comment (sanket143) [inline scripts/setup.js:45]: This much is not enough to show songs list sorted. This sorts `state.allFiles` but you also need to make sure whether the component which is rendering the data is using this or not.

In this case, `songsMap` will be empty and will throw an error for accessing "title" of undefined. For that, ` await extractAndStoreMetaTags()` should be before the sorting mechanism.

**Comment by sanket143** [comment]
You need to make following changes in https://github.com/AjaysinhRajput2802/MJam/blob/iss54/scripts/components/content.js#L46

```
-      console.log("Updated")
-      return Object.values(state.songsMap)
+      let songs = state.allFiles.map((item) => {
+        return state.songsMap[item]
+      })
+      return songs
```

**Comment by AjaysinhRajput2802** [comment]
@sanket143  made all changes requested by you and additional change to solve "title not defined" error.

File: scripts/state.js [hunk 5]
```diff
@@ -87,8 +87,9 @@ const state = new Vue({
         this.nowplaying.completion = 0
         clearInterval(this.nowplaying.tracker)
       })
-
-      this.nowplaying.song = this.songsMap[sources[0]]
+      if(Array.isArray(sources)){
+        this.nowplaying.song = this.songsMap[sources[0]]
+      }
```
Comment (sanket143) [inline scripts/state.js:92]: what is this for?

Comment (AjaysinhRajput2802) [reply scripts/state.js:92]: If we don't add this then we will get error of title not defined.

Comment (AjaysinhRajput2802) [reply scripts/state.js:92]: when sources is not array then we can't access sources[0], that's why we are getting error like "title of undefined".

Comment (AjaysinhRajput2802) [reply scripts/state.js:92]: @sanket143 Thus I made this change to solve that error.

Comment (AjaysinhRajput2802) [reply scripts/state.js]: Sorry @sanket143 this change is for issue #41 , I apologize for it and removed it from this PR

**Comment by sanket143** [comment]
@AjaysinhRajput2802  solve lint errors

**Review by sanket143 — APPROVED** [review]

**Comment by sanket143** [comment]
give @AjaysinhRajput2802 75 points

**Pull request merged**